blockstack_client/proxy.py

```diff
@@ -483,83 +915,258 @@ def get_consensus_range(block_id_start, block_id_end, proxy=None):
     """
     proxy = get_default_proxy() if proxy is None else proxy
 
-    resp = proxy.get_consensus_range(block_id_start, block_id_end)
+    ch_range = get_consensus_hashes(range(block_id_start, block_id_end + 1), proxy=proxy)
+    if 'error' in ch_range:
+        return ch_range
+
+    # verify that all blocks are included
+    for i in range(block_id_start, block_id_end + 1):
+        if i not in ch_range:
+            return {'error': 'Missing consensus hashes'}
+
+    return ch_range
+
+
+def get_name_blockchain_history(name, start_block, end_block, proxy=None):
+    """
+    Get the name's historical blockchain records.
+    Returns the list of states the name has been in on success
+    Returns {'error': ...} on error
+    """
+    schema_per_block = {
+        'type': 'array',
+        'items': {
+            'type': 'object',
+            'properties': NAMEOP_SCHEMA_PROPERTIES,
+            'required': NAMEOP_SCHEMA_REQUIRED
+        }
+    }
+
+    proxy = get_default_proxy() if proxy is None else proxy
+
+    resp = {}
+    try:
+        resp = proxy.get_name_blockchain_history(name, start_block, end_block)
+
+        for block_height in resp:
+            block_height_i = int(block_height)  # make sure this string encodes an int
+            block_resp = json_validate(schema_per_block, resp[block_height])
+            if json_is_error(block_resp):
+                return block_resp
+    except ValidationError as e:
+        resp = json_traceback(resp.get('error'))
 
     return resp
 
 
-def get_op_history_rows( name, proxy=None ):
+def get_op_history_rows(name, proxy=None):
```

Comment:
A lot of this functionality used to live in the server.  One of the heaviest-weight I/O-bound RPC requests a client can make is to ask for all of the name operations that occurred at a particular block height (`get_nameops_at`).  Since the server's internal representation of the name state is reminiscent of a version control system (i.e. each name record has a "head" state and a bunch of deltas that, when applied to it, restore it to a previous version of itself), calculating a previous name state required performing a bunch of disk queries.  This wasn't a problem in 0.13 since the entire DB lived in RAM.  But in 0.14, this became a bottleneck--the old `get_nameops_at` RPC call could take a long time to complete and, due to the black magic that powers Python's garbage collector, could sometimes run out of memory.

This method and a few of them below it are used to implement a more scalable version of `get_nameops_at`.  Instead of making a single RPC call, the client does the following:

* get all the names that were affected at the given block height (which can be done in batches)
* for each name, get the list of deltas (it's "history", which again can be done in batches)
* merge the name's head and deltas to calculate the name operation at that block height